File: lectures/lecture-06-why-initialization-needs-its-own-phase/exercises/exercise-02-readiness-gate/solution/typescript/main.ts
// readiness-gate exercise, TypeScript solution.
//
// Turns a set of readiness check results into a tiered verdict and exit
// code: blockers stop a session from starting; advice does not, but must
// stay visible. Contract: ../../SPEC.md.

import { readFileSync } from "node:fs";
import { pathToFileURL } from "node:url";

const EXIT_READY = 0;
const EXIT_BLOCKED = 1;
const EXIT_ADVICE = 3;

interface Check {
  readonly id: string;
  readonly severity: "blocker" | "advice";
  readonly passed: boolean;
}

export function gate(checks: readonly Check[]): [Record<string, unknown>, number] {
  const blockersFailed = checks
    .filter((c) => c.severity === "blocker" && !c.passed)
    .map((c) => c.id);
  const adviceFailed = checks
    .filter((c) => c.severity === "advice" && !c.passed)
    .map((c) => c.id);
  let verdict: string;
  let code: number;
  if (blockersFailed.length > 0) {
    verdict = "blocked";
    code = EXIT_BLOCKED;
  } else if (adviceFailed.length > 0) {
    verdict = "ready-with-advice";
    code = EXIT_ADVICE;
  } else {
    verdict = "ready";
    code = EXIT_READY;
  }
  return [
    { blockers_failed: blockersFailed, advice_failed: adviceFailed, verdict },
    code,
  ];
}

function main(argv: readonly string[]): number {
  const path = argv[2];
  if (!path || argv.length !== 3) {
    console.error("usage: main.ts <check-results.json>");
    return 2;
  }
  let data: { checks: Check[] };
  try {
    data = JSON.parse(readFileSync(path, "utf8")) as { checks: Check[] };
  } catch (error) {
    console.error(`error: cannot read results: ${String(error)}`);
    return 2;
  }
  const [report, code] = gate(data.checks);
  process.stdout.write(JSON.stringify(report, null, 2) + "\n");
  return code;
}

if (import.meta.url === pathToFileURL(process.argv[1] ?? "").href) {
  process.exit(main(process.argv));
}

File: lectures/lecture-06-why-initialization-needs-its-own-phase/exercises/exercise-02-readiness-gate/solution/python/main.py
"""readiness-gate exercise, Python solution.

Turns a set of readiness check results into a tiered verdict and exit
code: blockers stop a session from starting; advice does not, but must
stay visible. Contract: ../../SPEC.md.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

EXIT_READY = 0
EXIT_BLOCKED = 1
EXIT_ADVICE = 3


def gate(checks: list[dict]) -> tuple[dict, int]:
    blockers_failed = [c["id"] for c in checks if c["severity"] == "blocker" and not c["passed"]]
    advice_failed = [c["id"] for c in checks if c["severity"] == "advice" and not c["passed"]]
    if blockers_failed:
        verdict, code = "blocked", EXIT_BLOCKED
    elif advice_failed:
        verdict, code = "ready-with-advice", EXIT_ADVICE
    else:
        verdict, code = "ready", EXIT_READY
    report = {
        "blockers_failed": blockers_failed,
        "advice_failed": advice_failed,
        "verdict": verdict,
    }
    return report, code


def main(argv: list[str]) -> int:
    if len(argv) != 2:
        print("usage: main.py <check-results.json>", file=sys.stderr)
        return 2
    try:
        data = json.loads(Path(argv[1]).read_text(encoding="utf-8"))
    except OSError as error:
        print(f"error: cannot read results: {error}", file=sys.stderr)
        return 2
    report, code = gate(data["checks"])
    print(json.dumps(report, indent=2))
    return code


if __name__ == "__main__":
    sys.exit(main(sys.argv))
